### ISSUE TITLE
`juliaup gc` should remove `compiled` and `environments` folders that are no longer needed

### ISSUE THREAD

**Issue by numbermaniac** [JuliaLang/juliaup#1039]
I had Julia `release` 1.10.5 and `beta` 1.11rc3 both installed using Juliaup. I used `juliaup remove release` to remove the 1.10.5 installation, and then afterwards, used `juliaup gc`. To my surprise, the `.julia/compiled/v1.10/` and `.julia/environments/v1.10/` folders remained, which is significant because the first one was using 700MB of storage. I would've expected them to be deleted since they're no longer needed.

It would be great if `juliaup gc` could detect when such folders are no longer needed and remove them accordingly. Thank you!

(If it helps: Juliaup 1.17.4, macOS 14.6.1)

**Comment by davidanthoff** [comment]
My gut feeling is that we can/should do something about the `.julia/compiled` folder, but not `.julia/environments`. The `compiled` folder is really just a cache/optimization, so deleting that seems fine. But the `environments` folder is really user-controlled/created data, and I think we should just never delete anything there.

**Comment by KristofferC** [comment]
Should imo be a part of Pkg.gc since that already handles artifacts and packages.

**Comment by davidanthoff** [comment]
But `Pkg.gc` presumably doesn't know which Julia versions are installed, right?

**Comment by numbermaniac** [comment]
Yeah - I assume a `Pkg` in one Julia installation wouldn't know what other Julia installations you have, so it would need to be managed by juliaup.

**Comment by StefanKarpinski** [comment]
Could potentially be a two-step thing where `juliaup gc` writes a file that marks something as potentially unused and then `Pkg.gc` uses that to figure out what can be deleted. But seems kind of slow to kick in then...

**Comment by davidanthoff** [comment]
We could also add a secret command to `juliaup` that Pkg.jl could use to get a list of all currently installed Julia versions. `Pkg.gc` could try to spawn that command then, and if it gets something useful back use that information to decide which folders to delete.

I think in general we are probably reaching a point where Juliaup and Pkg.jl should be more integrated and talk to each other.

**Comment by StefanKarpinski** [comment]
> I think in general we are probably reaching a point where Juliaup and Pkg.jl should be more integrated and talk to each other.

Yes, it does seem like we're at that point.

**Comment by danvinci** [comment]
> > I think in general we are probably reaching a point where Juliaup and Pkg.jl should be more integrated and talk to each other.
> 
> Yes, it does seem like we're at that point.

Seems that Juliaup exposes the installed versions via `api getconfig1` and it emits JSON.

Pkg has no JSON dependency, but they both have TOML - so an option would be to parametrize `getconfig1` to also output TOML, and have Pkg.gc() consume that to drop compiled/vX.Y for minors that are no longer installed.

What you think?

Happy to take a stab at it if it makes sense.

**Comment by KristofferC** [comment]
If I have a locally compiled v1.12 I don't want these to be removed. I'm not sure it is right to think that just because juliaup is installed means that juliaup knows about all julias on a system.

**Comment by danvinci** [comment]
Riight 🤦🏼 

Alternative: Pkg.gc() could use the last-used timestamp in .ji files.
Base touches the cache on every load - unless files are read-only for some reason - [here](https://github.com/JuliaLang/julia/blob/51865f0c283c3eacde2429df8e4c2900eb121587/base/loading.jl#L2036-L2037) and [here](https://github.com/JuliaLang/julia/blob/51865f0c283c3eacde2429df8e4c2900eb121587/base/loading.jl#L2258) - so the used Julia version keeps its caches fresh.

More aligned to JuliaLang/Pkg.jl#2684 

Yet to be tested though, I might be missing the bigger picture or ways the timestamp might not be reliable. 
I could try and draft a PR on the Pkg repo if the approach makes sense.
